crates/sui-data-ingestion/src/workers/archival.rs

```diff
@@ -138,67 +112,69 @@ impl ArchivalWorker {
-            checkpoint
-                .transactions
-                .iter()
-                .map(|t| ExecutionData::new(t.transaction.clone(), t.effects.clone())),
-        );
-        let contents_blob = Blob::encode(&full_checkpoint_contents, BlobEncoding::Bcs)?;
-        let blob_size = contents_blob.size();
-        let summary_blob = Blob::encode(&checkpoint.checkpoint_summary, BlobEncoding::Bcs)?;
-
-        if !state.buffer.is_empty()
-            && (((state.buffer.len() + blob_size) > self.commit_file_size)
-                || state.epoch != epoch
-                || checkpoint.checkpoint_summary.timestamp_ms
-                    > (self.commit_duration_ms + state.last_commit_ms))
-        {
-            self.upload(&state).await?;
-            state.epoch = epoch;
-            state.checkpoint_range = sequence_number..sequence_number;
-            state.buffer = vec![];
-            state.summary_buffer = vec![];
-            state.last_commit_ms = checkpoint.checkpoint_summary.timestamp_ms;
-            state.should_update_progress = true;
+        let mut summary_buffer = vec![];
+        let mut buffer = vec![];
+        let first_checkpoint = &batch[0];
+        let epoch = first_checkpoint.checkpoint_summary.epoch;
+        let start_checkpoint = first_checkpoint.checkpoint_summary.sequence_number;
+        let mut last_checkpoint = start_checkpoint;
+        for checkpoint in batch {
+            let full_checkpoint_contents = FullCheckpointContents::from_contents_and_execution_data(
+                checkpoint.checkpoint_contents.clone(),
+                checkpoint
+                    .transactions
+                    .iter()
+                    .map(|t| ExecutionData::new(t.transaction.clone(), t.effects.clone())),
+            );
+            let contents_blob = Blob::encode(&full_checkpoint_contents, BlobEncoding::Bcs)?;
+            let summary_blob = Blob::encode(&checkpoint.checkpoint_summary, BlobEncoding::Bcs)?;
+            contents_blob.write(&mut buffer)?;
+            summary_blob.write(&mut summary_buffer)?;
+            last_checkpoint += 1;
         }
-        contents_blob.write(&mut state.buffer)?;
-        summary_blob.write(&mut state.summary_buffer)?;
-        state.checkpoint_range.end += 1;
+        self.upload(
+            epoch,
+            start_checkpoint,
+            last_checkpoint,
+            summary_buffer,
+            buffer,
+        )
+        .await?;
         Ok(())
     }
 
-    async fn save_progress(
+    fn should_close_batch(
         &self,
-        sequence_number: CheckpointSequenceNumber,
-    ) -> Option<CheckpointSequenceNumber> {
-        let mut state = self.state.lock().await;
-        let should_update_progress = state.should_update_progress;
-        state.should_update_progress = false;
-        if should_update_progress && sequence_number > 0 {
-            Some(sequence_number - 1)
-        } else {
-            None
+        batch: &[CheckpointData],
```

Comment:
It would be nice if this could be a reference the the Arc wrapped CheckpointData's vs needing to clone them all...but this is workflow dependent...Maybe not worth iterating on right now though...